src/main/java/com/arpnetworking/metrics/impl/ApacheHttpSink.java

```diff
@@ -333,87 +370,53 @@ private void extractTimestamp(final Event event, final String key, final LongCon
-                baseUnit = (BaseUnit) unit;
-                baseScale = null;
-            } else {
-                return null;
-            }
-
-            final ClientV2.Unit.Builder builder = ClientV2.Unit.newBuilder();
-            try {
-                builder.setType(ClientV2.Unit.Type.Value.valueOf(baseUnit.name()));
-            } catch (final IllegalArgumentException e) {
-                _dispatchErrorLogger.getLogger().warn(
-                        String.format("Dropping unsupported unit %s", baseUnit.name()));
-                builder.setType(ClientV2.Unit.Type.Value.UNKNOWN);
-            }
-            if (baseScale != null) {
-                try {
-                    builder.setScale(ClientV2.Unit.Scale.Value.valueOf(baseScale.name()));
-                } catch (final IllegalArgumentException e) {
-                    _dispatchErrorLogger.getLogger().warn(
-                            String.format("Dropping unsupported scale %s", baseScale.name()));
-                    builder.setScale(ClientV2.Unit.Scale.Value.UNKNOWN);
-                }
-            } else {
-                builder.setScale(ClientV2.Unit.Scale.Value.UNIT);
+        private ClientV3.NumericalData buildNumericalData(final List<Quantity> samples, @Nullable final AggregatedData aggregatedData) {
+            final ClientV3.NumericalData.Builder builder = ClientV3.NumericalData.newBuilder();
+            builder.addAllSamples(samples.stream().map(q -> q.getValue().doubleValue()).collect(Collectors.toList()));
+            if (aggregatedData instanceof AugmentedHistogram) {
+                final AugmentedHistogram augmentedHistogram = (AugmentedHistogram) aggregatedData;
+                final int precision = augmentedHistogram.getPrecision();
+                final long truncateMask = BASE_MASK >> precision;
+                final long packMask = (1 << (precision + EXPONENT_BITS + 1)) - 1;
+                builder.setStatistics(
+                        ClientV3.AugmentedHistogram.newBuilder()
+                                .setPrecision(augmentedHistogram.getPrecision())
+                                .setMin(augmentedHistogram.getMin())
+                                .setMax(augmentedHistogram.getMax())
+                                .setSum(augmentedHistogram.getSum())
+                                .addAllEntries(
+                                        augmentedHistogram.getHistogram()
+                                                .entrySet()
+                                                .stream()
+                                                .map(entry -> {
+                                                    final long truncatedKey = Double.doubleToRawLongBits(entry.getKey()) & truncateMask;
+                                                    final long shiftedKey = truncatedKey >> (MANTISSA_BITS - precision);
+                                                    final long packedKey = shiftedKey & packMask;
```

Comment:
That's a fair question.

The `AugmentedHistogram` type from Metrics Client Java is intended as a serialization independent type. The keys in this map should already be truncated. I have clarified this in the Javadoc:

https://github.com/ArpNetworking/metrics-client-java/pull/53

The code here in the Metrics Apache Http Sink Extra should serialize the histogram keys into Protobuf. The process here involves converting the truncated double key into a packed long that is optimized for var-int encoding.

While not broken, the truncation here is redundant based on the `AugmentedHistogram` contract.

The code to pack the key into this serialization format should no be `AugmentedHistogram` because that representation should be serialization format independent.

Does that answer your question?

Reply:
Yup, if this packing is serialization-specific then it's fine.

I'd like this documented somewhere with the protocol (`SparseHistogramEntry`) since that wasn't immediately apparent to me that using AugmentedHistogram in this way is specified by the interface and not an implementation detail.